core/src/main/scala/org/apache/spark/api/java/JavaUtils.scala

```diff
@@ -43,10 +43,17 @@ private[spark] object JavaUtils {
 
     override def size: Int = underlying.size
 
-    override def get(key: AnyRef): B = try {
-      underlying.getOrElse(key.asInstanceOf[A], null.asInstanceOf[B])
-    } catch {
-      case ex: ClassCastException => null.asInstanceOf[B]
+    // Delegate to implementation because AbstractMap implementation iterates over whole key set
+    override def containsKey(key: AnyRef): Boolean = {
+      underlying.contains(key.asInstanceOf[A])
```

Comment:
Really, this should return `false` if the key isn't an `A`. This will throw an exception now. It should be prefixed with `key.isInstanceOf[A] && ...`

Reply:
I thought it should throw exception, however there is a test showing that it's fine...

Reply:
Yeah for historical reasons the signature has a key of type AnyRef instead of A (pre-generics compatibility) and so the behavior for a key of the wrong class can go either way: https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#optional-restrictions We should probably just keep the behavior.

Reply:
do we mean we can just write `underlying.contains(key)`?

Reply:
No, because Scala Maps do require their key type in contains() and get(). https://www.scala-lang.org/api/current/scala/collection/Map.html . This is indeed the difference between the APIs that we need to bridge.

Reply:
@cloud-fan @srowen 
So, let's have decided what we want for containsKey() here first, before i will do next commit -)
Because suggested (key.isInstanceOf[A] && underlying.contains(key.asInstanceOf[A]))
will not work, it cause compile-time error with "abstract type A is unchecked since it is eliminated by erasure" message.

Reply:
Neveminde, maybe it's my local issues, i will commit suggested chagnes, lets look at jenkins afterall.